From 5ef87c82df86c6a23da330683d5d356e117458e4 Mon Sep 17 00:00:00 2001
From: Simon Josefsson <simon@josefsson.org>
Date: Mon, 03 Aug 2009 21:40:23 +0000
Subject: Fix expected output, a cert have expired.

---
diff --git a/tests/chainverify.c b/tests/chainverify.c
index 73a3a85..f394a28 100644
--- a/tests/chainverify.c
+++ b/tests/chainverify.c
@@ -690,8 +690,12 @@ static struct
     0, GNUTLS_CERT_SIGNER_NOT_CA | GNUTLS_CERT_INVALID },
   { "hbci v1 fail", hbci_chain, &hbci_chain[2],
     0, GNUTLS_CERT_SIGNER_NOT_CA | GNUTLS_CERT_INVALID},
+  { "hbci v1 ok expired", hbci_chain, &hbci_chain[2],
+    GNUTLS_VERIFY_ALLOW_X509_V1_CA_CRT,
+    GNUTLS_CERT_EXPIRED | GNUTLS_CERT_INVALID },
   { "hbci v1 ok", hbci_chain, &hbci_chain[2],
-    GNUTLS_VERIFY_ALLOW_X509_V1_CA_CRT, 0 },
+    GNUTLS_VERIFY_DISABLE_TIME_CHECKS | GNUTLS_VERIFY_ALLOW_X509_V1_CA_CRT,
+    0 },
   { "rsa-md5 fail", mayfirst_chain, &mayfirst_chain[1],
     0, GNUTLS_CERT_INSECURE_ALGORITHM | GNUTLS_CERT_INVALID },
   { "rsa-md5 not ok", mayfirst_chain, &mayfirst_chain[1],
--
cgit v0.8.2.1
