diff -NurbB gnutls-2.7.10.ginol/libextra/gnutls-extra.pc.in gnutls-2.7.10/libextra/gnutls-extra.pc.in
--- gnutls-2.7.10.ginol/libextra/gnutls-extra.pc.in	2009-05-11 18:15:43.000000000 +0200
+++ gnutls-2.7.10/libextra/gnutls-extra.pc.in	2009-05-14 19:33:54.000000000 +0200
@@ -21,5 +21,5 @@
 Requires: gnutls
 Version: @VERSION@
 Libs: -L${libdir} -lgnutls-extra
-Libs.private: @LIBGNUTLS_EXTRA_LIBS@
+Libs.private: @LIBGNUTLS_EXTRA_LIBS@ @LTLIBTASN1@
 Cflags: -I${includedir}
