From 6fffdf51ce25f68c8b7fb370b222d6f923f650b3 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 16 Apr 2011 22:46:56 +0200
Subject: [PATCH] Restored HMAC-MD5 for compatibility. Although considered weak, several sites require it for connection. It is enabled for "NORMAL" and "PERFORMANCE" priority strings.
Bug-Debian: http://bugs.debian.org/623001

diff --git a/lib/gnutls_priority.c b/lib/gnutls_priority.c
index 2f35a84..8f59207 100644
--- a/lib/gnutls_priority.c
+++ b/lib/gnutls_priority.c
@@ -355,9 +355,10 @@ static const int sign_priority_secure256[] = {
   0
 };
 
-static const int mac_priority_performance[] = {
+static const int mac_priority_normal[] = {
   GNUTLS_MAC_SHA1,
   GNUTLS_MAC_SHA256,
+  GNUTLS_MAC_MD5,
   0
 };
 
@@ -573,7 +574,7 @@ gnutls_priority_init (gnutls_priority_t * priority_cache,
           _set_priority (&(*priority_cache)->cipher,
                          cipher_priority_performance);
           _set_priority (&(*priority_cache)->kx, kx_priority_performance);
-          _set_priority (&(*priority_cache)->mac, mac_priority_performance);
+          _set_priority (&(*priority_cache)->mac, mac_priority_normal);
           _set_priority (&(*priority_cache)->sign_algo,
                          sign_priority_default);
         }
@@ -581,7 +582,7 @@ gnutls_priority_init (gnutls_priority_t * priority_cache,
         {
           _set_priority (&(*priority_cache)->cipher, cipher_priority_normal);
           _set_priority (&(*priority_cache)->kx, kx_priority_secure);
-          _set_priority (&(*priority_cache)->mac, mac_priority_secure);
+          _set_priority (&(*priority_cache)->mac, mac_priority_normal);
           _set_priority (&(*priority_cache)->sign_algo,
                          sign_priority_default);
         }
-- 
1.7.2.5

