diff -NurbBp gnutls26-2.8.4/tests/chainverify.c gnutls-2.8.4.new/tests/chainverify.c
--- gnutls26-2.8.4/tests/chainverify.c	2009-08-07 01:22:23.000000000 +0200
+++ gnutls-2.8.4.new/tests/chainverify.c	2009-11-01 13:12:06.000000000 +0100
@@ -32,6 +32,23 @@
 #include <gnutls/gnutls.h>
 #include <gnutls/x509.h>
 
+
+/* GnuTLS internally calls time() to find out the current time when
+   verifying certificates.  To avoid a time bomb, we hard code the
+   current time.  This should work fine on systems where the library
+   call to time is resolved at run-time.  */
+time_t
+time (time_t *t)
+{
+  time_t then = 1255816800;
+
+  if (t)
+    *t = then;
+
+  return then;
+}
+
+
 /* *INDENT-OFF* */
 
 /* Triggers incorrect verification success on older versions */
