From 7daf2300d81dbeeee870299ba3bf7d1ebe3a0d12 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 16 Dec 2011 05:02:45 +0100
Subject: [PATCH] correctly set the odd bits.

---
 lib/nettle/mpi.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/nettle/mpi.c b/lib/nettle/mpi.c
index 0ebdcf4..f1342ec 100644
--- a/lib/nettle/mpi.c
+++ b/lib/nettle/mpi.c
@@ -474,7 +474,7 @@ gen_group (mpz_t * prime, mpz_t * generator, unsigned int nbits, unsigned int *q
 
       nettle_mpz_set_str_256_u (w, w_bytes, buffer);
       /* always odd */
-      mpz_setbit (q, 0);
+      mpz_setbit (w, 0);
 
       ret = mpz_probab_prime_p (w, PRIME_CHECK_PARAM);
       if (ret > 0)
@@ -500,7 +500,7 @@ gen_group (mpz_t * prime, mpz_t * generator, unsigned int nbits, unsigned int *q
 
       nettle_mpz_set_str_256_u (q, q_bytes, buffer);
       /* always odd */
-      mpz_setbit (w, 0);
+      mpz_setbit (q, 0);
 
       ret = mpz_probab_prime_p (q, PRIME_CHECK_PARAM);
       if (ret == 0)
-- 
1.7.2.5

