Description: Link against debian libopts.
Author: Andreas Metzler <ametzler@debian.org>
Last-Update: 2013-08-24

--- gnutls28-3.2.3.orig/configure
+++ gnutls28-3.2.3/configure
@@ -52951,10 +52951,6 @@ else
 fi
 
 
-enable_local_libopts=yes
-
-NEED_LIBOPTS_DIR=true
-
 
   ac_header_dirent=no
 for ac_hdr in dirent.h sys/ndir.h sys/dir.h ndir.h; do
@@ -53917,9 +53913,6 @@ _ACEOF
 
   LIBOPTS_BUILD_BLOCKED=''
 
-
-  NEED_LIBOPTS_DIR=''
-
   # Check whether --enable-local-libopts was given.
 if test "${enable_local_libopts+set}" = set; then :
   enableval=$enable_local_libopts;
@@ -54517,10 +54510,10 @@ $as_echo "#define NO_OPTIONAL_OPT_ARGS 1
 
 
 
-    ac_config_files="$ac_config_files src/libopts/Makefile"
 
 
           fi    # end of AC_DEFUN of LIBOPTS_CHECK
+ac_config_files="$ac_config_files src/libopts/Makefile"
 
 
 ac_fn_c_check_type "$LINENO" "ssize_t" "ac_cv_type_ssize_t" "
