From 9c66a020d7e424c19bb4eb1ba33d9603a0f03f45 Mon Sep 17 00:00:00 2001
From: Matt Whitlock <matt@whitlock.name>
Date: Tue, 16 Jul 2013 07:24:10 -0400
Subject: [PATCH 3/4] avoid leaking a buffer element when _gnutls_stream_read
 returns 0

---
 lib/gnutls_buffers.c |    3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/gnutls_buffers.c b/lib/gnutls_buffers.c
index 9e4735e..3b281f7 100644
--- a/lib/gnutls_buffers.c
+++ b/lib/gnutls_buffers.c
@@ -355,6 +355,9 @@ finish:
   _gnutls_read_log ("READ: read %d bytes from %p\n",
                         (int) (size - left), fd);
 
+  if (size - left == 0)
+    _mbuffer_xfree(bufel);
+
   return (size - left);
   
 cleanup:
-- 
1.7.10.4

