Description: Disable parallel build in Gguile subdir.
Author: Andreas Metzler <ametzler@debian.org>
Origin: vendor
Last-Update: 2013-09-09

--- gnutls28-3.2.4.orig/guile/Makefile.in
+++ gnutls28-3.2.4/guile/Makefile.in
@@ -1551,3 +1551,4 @@ uninstall-am:
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
 .NOEXPORT:
+.NOTPARALLEL:
--- gnutls28-3.2.4.orig/guile/modules/Makefile.in
+++ gnutls28-3.2.4/guile/modules/Makefile.in
@@ -1477,3 +1477,4 @@ uninstall-am: uninstall-nobase_dist_guil
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
 .NOEXPORT:
+.NOTPARALLEL:
--- gnutls28-3.2.4.orig/guile/src/Makefile.in
+++ gnutls28-3.2.4/guile/src/Makefile.in
@@ -1693,3 +1693,4 @@ built-sources: $(BUILT_SOURCES)
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
 .NOEXPORT:
+.NOTPARALLEL:
--- gnutls28-3.2.4.orig/guile/tests/Makefile.in
+++ gnutls28-3.2.4/guile/tests/Makefile.in
@@ -1827,3 +1827,4 @@ uninstall-am:
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
 .NOEXPORT:
+.NOTPARALLEL:
