From 43e14e8a5cef27f319316bb7ff0de8b86272ac99 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 2 Sep 2012 10:36:12 +0200
Subject: [PATCH 1/4] use errno_to_gerr() in _gnutls_io_check_recv().

---
 lib/gnutls_buffers.c |   10 +++-------
 1 file changed, 3 insertions(+), 7 deletions(-)

diff --git a/lib/gnutls_buffers.c b/lib/gnutls_buffers.c
index c714519..44b09cb 100644
--- a/lib/gnutls_buffers.c
+++ b/lib/gnutls_buffers.c
@@ -663,13 +663,9 @@ _gnutls_io_check_recv (gnutls_session_t session, unsigned int ms)
   ret = session->internals.pull_timeout_func(fd, ms);
 
   err = get_errno (session);
-  if (ret == -1 && err == EINTR)
-    return GNUTLS_E_INTERRUPTED;
-  else if (ret == -1 && err == EAGAIN)
-    return GNUTLS_E_AGAIN;
-  else if (ret == -1)
-    return gnutls_assert_val(GNUTLS_E_PULL_ERROR);
-  
+  if (ret == -1)                 
+    return errno_to_gerr(err);
+
   if (ret > 0)
     return 0;
   else return GNUTLS_E_TIMEDOUT;
-- 
1.7.10.4

