From ef9d0e356c191db966b7eb5ed8e55e7155b029c3 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 2 Sep 2012 10:59:16 +0200
Subject: [PATCH 3/4] dump the errno received by select

---
 lib/gnutls_buffers.c |    8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/lib/gnutls_buffers.c b/lib/gnutls_buffers.c
index 546293f..ea7a77a 100644
--- a/lib/gnutls_buffers.c
+++ b/lib/gnutls_buffers.c
@@ -657,8 +657,12 @@ _gnutls_io_check_recv (gnutls_session_t session, unsigned int ms)
   ret = session->internals.pull_timeout_func(fd, ms);
 
   err = get_errno (session);
-  if (ret == -1)                 
-    return errno_to_gerr(err);
+  if (ret == -1)
+    {
+      _gnutls_read_log ("READ_TIMEOUT: %d returned from %p, errno=%d (timeout: %u)\n",
+			(int) ret, fd, err, ms);
+      return errno_to_gerr(err);
+    }
 
   if (ret > 0)
     return 0;
-- 
1.7.10.4

