From 9b75c263502f85911bf5d07eb5cbc1dc57139a26 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Thu, 30 Apr 2015 18:48:57 +0200
Subject: [PATCH 19/19] doc: updated gnutls_dtls_set_timeouts

---
 lib/gnutls_dtls.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/lib/gnutls_dtls.c b/lib/gnutls_dtls.c
index ab02721..37fcb62 100644
--- a/lib/gnutls_dtls.c
+++ b/lib/gnutls_dtls.c
@@ -550,9 +550,6 @@ int _dtls_record_check(struct record_parameters_st *rp, uint64 * _seq)
  * The DTLS protocol recommends the values of 1 sec and 60 seconds
  * respectively.
  *
- * If the retransmission timeout is zero then the handshake will operate
- * in a non-blocking way, i.e., return %GNUTLS_E_AGAIN.
- *
  * To disable retransmissions set a @retrans_timeout larger than the @total_timeout.
  *
  * Since: 3.0
-- 
2.1.4

