From d091c669dae2c3fad8c3cc6a19ebff5d212a4b54 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@redhat.com>
Date: Tue, 21 Apr 2015 10:40:10 +0200
Subject: [PATCH 05/19] doc update

---
 NEWS | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/NEWS b/NEWS
index 011d0db..1897780 100644
--- a/NEWS
+++ b/NEWS
@@ -3,6 +3,19 @@ Copyright (C) 2000-2015 Free Software Foundation, Inc.
 Copyright (C) 2013, 2014 Nikos Mavrogiannopoulos
 See the end for copying conditions.
 
+* Version 3.3.15 (unreleased)
+
+** libgnutls: gnutls_certificate_get_ours: will return the certificate even
+if a callback was used to send it.
+
+** Check for invalid length in the X.509 version field. Without the check
+certificates with invalid length would be detected as having an arbitrary
+version. Reported by Hanno Böck.
+
+** API and ABI modifications:
+No changes since last version.
+
+
 * Version 3.3.14 (released 2015-03-30)
 
 ** libgnutls: When retrieving OCTET STRINGS from PKCS #12 ContentInfo
-- 
2.1.4

