From b9924dafb6262563077ea7db2c98f168b062da20 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@redhat.com>
Date: Thu, 30 Apr 2015 16:28:24 +0200
Subject: [PATCH 18/19] gnutls_handshake_set_timeout will properly work with
 DTLS

---
 lib/gnutls_handshake.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/gnutls_handshake.c b/lib/gnutls_handshake.c
index 2b8a291..500d252 100644
--- a/lib/gnutls_handshake.c
+++ b/lib/gnutls_handshake.c
@@ -2597,6 +2597,11 @@ int gnutls_handshake(gnutls_session_t session)
 void
 gnutls_handshake_set_timeout(gnutls_session_t session, unsigned int ms)
 {
+	if (IS_DTLS(session)) {
+		gnutls_dtls_set_timeouts(session, 1000, ms);
+		return;
+	}
+
 	if (ms == GNUTLS_DEFAULT_HANDSHAKE_TIMEOUT)
 		ms = 40 * 1000;
 	session->internals.handshake_timeout_ms = ms;
-- 
2.1.4

