From d9c4ce676d140a63ae0ebd719d67e85ea13eea6b Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 25 Apr 2015 21:24:17 +0200
Subject: [PATCH 11/19] doc update

---
 NEWS | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/NEWS b/NEWS
index 1897780..7d0eee4 100644
--- a/NEWS
+++ b/NEWS
@@ -8,7 +8,10 @@ See the end for copying conditions.
 ** libgnutls: gnutls_certificate_get_ours: will return the certificate even
 if a callback was used to send it.
 
-** Check for invalid length in the X.509 version field. Without the check
+** libgnutls: Fix for MD5 downgrade in TLS 1.2 signatures. Reported by
+Karthikeyan Bhargavan.
+
+** libgnutls: Check for invalid length in the X.509 version field. Without the check
 certificates with invalid length would be detected as having an arbitrary
 version. Reported by Hanno Böck.
 
-- 
2.1.4

