From a46eee667ae793aa9d178b1dd4708ddd72af8be1 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@redhat.com>
Date: Wed, 22 Apr 2015 13:41:17 +0200
Subject: [PATCH 06/19] fixed doc: reported by Anton Lavrentiev

---
 lib/gnutls_cert.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/gnutls_cert.c b/lib/gnutls_cert.c
index 3132204..42cd23f 100644
--- a/lib/gnutls_cert.c
+++ b/lib/gnutls_cert.c
@@ -391,7 +391,7 @@ void gnutls_certificate_server_set_retrieve_function
  * int (*callback)(gnutls_session_t, const gnutls_datum_t* req_ca_dn, int nreqs,
  * const gnutls_pk_algorithm_t* pk_algos, int pk_algos_length, gnutls_retr2_st* st);
  *
- * @req_ca_cert is only used in X.509 certificates.
+ * @req_ca_dn is only used in X.509 certificates.
  * Contains a list with the CA names that the server considers trusted.
  * Normally we should send a certificate that is signed
  * by one of these CAs. These names are DER encoded. To get a more
@@ -433,7 +433,7 @@ void gnutls_certificate_set_retrieve_function
  * const gnutls_pk_algorithm_t* pk_algos, int pk_algos_length, gnutls_pcert_st** pcert,
  * unsigned int *pcert_length, gnutls_privkey_t * pkey);
  *
- * @req_ca_cert is only used in X.509 certificates.
+ * @req_ca_dn is only used in X.509 certificates.
  * Contains a list with the CA names that the server considers trusted.
  * Normally we should send a certificate that is signed
  * by one of these CAs. These names are DER encoded. To get a more
@@ -442,7 +442,7 @@ void gnutls_certificate_set_retrieve_function
  * @pk_algos contains a list with server's acceptable signature algorithms.
  * The certificate returned should support the server's given algorithms.
  *
- * @pcert should contain a single certificate and public or a list of them.
+ * @pcert should contain a single certificate and public key or a list of them.
  *
  * @pcert_length is the size of the previous list.
  *
-- 
2.1.4

