From d9a3c7b6f646fffb16f281c8776a3ab394c64ade Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@redhat.com>
Date: Mon, 20 Apr 2015 16:25:21 +0200
Subject: [PATCH 03/19] ensure that the X.509 version number is one byte only

---
 lib/x509/x509.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/x509/x509.c b/lib/x509/x509.c
index eb59e19..db8b07d 100644
--- a/lib/x509/x509.c
+++ b/lib/x509/x509.c
@@ -744,7 +744,7 @@ int gnutls_x509_crt_get_version(gnutls_x509_crt_t cert)
 		return _gnutls_asn2err(result);
 	}
 
-	if (len == 0)
+	if (len != 1)
 		return gnutls_assert_val(GNUTLS_E_CERTIFICATE_ERROR);
 
 	return (int) version[0] + 1;
-- 
2.1.4

