From d8f455b832ddebf5450899380825db93e91199af Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Wed, 10 Jun 2015 21:53:05 +0200
Subject: [PATCH] Corrected camellia256 set key in nettle3 compat mode

---
 lib/nettle/cipher.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/nettle/cipher.c b/lib/nettle/cipher.c
index 389836f..5c90d49 100644
--- a/lib/nettle/cipher.c
+++ b/lib/nettle/cipher.c
@@ -512,10 +512,10 @@ wrap_nettle_cipher_setkey(void *_ctx, const void *key, size_t keysize)
 	case GNUTLS_CIPHER_CAMELLIA_256_CBC:
 #ifdef USE_NETTLE3
 		if (ctx->enc)
-			camellia192_set_encrypt_key(ctx->ctx_ptr,
+			camellia256_set_encrypt_key(ctx->ctx_ptr,
 						 key);
 		else
-			camellia192_set_decrypt_key(ctx->ctx_ptr,
+			camellia256_set_decrypt_key(ctx->ctx_ptr,
 						 key);
 		break;
 #else
-- 
2.1.4

